### PR TITLE
Improved identifier registry macro

### PR DESCRIPTION
**Pull request by zaaarf**
The macro now also declares the registry and its initial value, sparing us a lot of constructors.

I also took the liberty of slightly tweaking the existing stuff to reuse more code.

Left it as a draft because it shouldn't be merged with pending stuff, to spare us some rebases.

### CODE REVIEW

**Comment by zaaarf** [comment]
I'll experiment a bit to find what's causing the MSVC problem, will squash once I have it working, don't worry

**Comment by Spartan322** [comment]
Needs to be updated for events now.

**Comment by zaaarf** [comment]
> Needs to be updated for events now.

Yeah, though I'll wait until my other stuff is merged too, it'll need to be updated for decisions too.

**Comment by Hop311** [comment]
Why not just leave the registry type outside the macro, similar to `PROPERTY`?
```c++
NamedInstanceRegistry<Element, UIManager const&> IDENTIFIER_REGISTRY(window_element);
```

File: src/openvic-simulation/types/IdentifierRegistry.hpp [hunk 1]
```diff
@@ -315,98 +315,85 @@ namespace OpenVic {
...
+#define IDENTIFIER_REGISTRY_FULL_CUSTOM(registry_type, singular, plural, registry, debug_name, index_offset) \
+	registry_type plural { #debug_name };\
+public: \
 	void lock_##plural() { \
 		registry.lock(); \
 	} \
-	bool plural##_are_locked() const { \
+	bool plural##_are_locked() \
+	const \
+	{ \
```
Comment (Hop311) [inline src/openvic-simulation/types/IdentifierRegistry.hpp:337]:
```suggestion
	bool plural##_are_locked() const { \
```
Also can you move these onto the same line

**Comment by zaaarf** [comment]
> Why not just leave the registry type outside the macro, similar to `PROPERTY`?
> 
> ```c++
> NamedInstanceRegistry<Element, UIManager const&> IDENTIFIER_REGISTRY(window_element);
> ```

Because I didn't think of it. I'll fix it now.

File: src/openvic-simulation/interface/GUI.hpp [hunk 2]
```diff
@@ -40,15 +40,14 @@ namespace OpenVic::GUI {
...
+		NamedInstanceRegistry<Element, UIManager const&> IDENTIFIER_REGISTRY(scene_element)
 
 	protected:
-		Scene();
+		Scene() = default;
 
 		bool _fill_key_map(NodeTools::key_map_t& key_map, UIManager const& ui_manager) override;
 
 	public:
-		Scene(Scene&&) = default;
```
Comment (Hop311) [inline src/openvic-simulation/interface/GUI.hpp:51]: Couple small things in this file:
- can you add back the default move constructor here,
- and add semicolons to the ends of the `scene_element` and `window_element` registry declarations,
- and add back the newline back at the end of the file.

**Review by Hop311 — COMMENTED** [review]
Great work, this clears up so much boilerplate code, thank you

File: src/openvic-simulation/map/Province.hpp [hunk 3]
```diff
@@ -80,7 +80,7 @@ namespace OpenVic {
 		Crime const* PROPERTY_RW(crime);
 		// TODO - change this into a factory-like structure
 		Good const* PROPERTY(rgo);
-		IdentifierRegistry<BuildingInstance> buildings;
+		IdentifierRegistry<BuildingInstance> IDENTIFIER_REGISTRY(building);
```
[the remark this answers is not in the document]
Comment (zaaarf) [reply src/openvic-simulation/map/Province.hpp:83]: Yeah, it was intentional. I preferred having the preprocessor repeating a bit of code to not having boilerplate.

File: src/openvic-simulation/pop/Pop.cpp [hunk 4]
```diff
@@ -40,8 +40,6 @@ PopType::PopType(
 	assert(merge_max_size >= 0);
 }
 
-PopManager::PopManager() : pop_types { "pop types" }, slave_sprite { 0 }, administrative_sprite { 0 } {}
```
[the remark this answers is not in the document]
Comment (zaaarf) [reply src/openvic-simulation/pop/Pop.cpp:43]: Oh, sorry. That was an oversight.

**Review by Hop311 — APPROVED** [review]

**Pull request merged**